### ISSUE TITLE
Database schema for sentiment analysis Group 5

### ISSUE THREAD

**Issue by kajo404** [Rostlab/JS16_ProjectA#80]
Hello Guys,

we need to store our sentiment data in your database and you get to write the API for it :)

please check out this [issue](https://github.com/Rostlab/JS16_ProjectD_Group5/issues/26) to get an idea of what we want to store. 
Here is a short summary: We need a schema that stores a list (1000+ elements) of a Date and 5 numbers for every character. 

Functionality we need:
- Add an element to list of a character 
- Get parts of a list for a character with a start and end date

I think that is it so far. Feel free to ask any questions you might have

**Comment by kordianbruck** [comment]
As written in the email, the API is already defined for characters to edit: https://got-api.bruck.me/doc/#api-Characters-EditCharacter

In your case, as the data set will be quite large, we might want to think about adding a separate collection. Can we please get a json representation of the data structure? Or comment on the following

```
[
  {
   character - reference to the character ID
   date - type would be date for this
   values: [int]
   },
   ....
]
```

Please provide us with proper naming for the structure.

**Comment by kajo404** [comment]
I think adding a seperate structure is the way to go as Group 4 will most likely need something like this aswell. We dont work with the character id, we use the names we get from your database. A json representation could like this:

```
{
'characterName' : <String>,
'date' : <Date>,   //in js date format
'posSum' : <int>,
'negSum' : <int>,
'posCount' : <int>,
'negCount' : <int>,
'nullCount' : <int>
}
```

**Comment by kordianbruck** [comment]
@kajo404 please check https://github.com/Rostlab/JS16_ProjectA/blob/master/app/models/characterSentiment.js

I'd rather use the _id instead of name field, as it is guaranteed unique and mongodb has an index set for it. Is that too much of a hassle? @sacdallago can you advise on this as well?

**Comment by togiberlin** [comment]
@kordianbruck lol, we pushed the same thing at the same time...i'll delete mine then

**Comment by kordianbruck** [comment]
I pushed mine first :P

**Comment by kajo404** [comment]
That Schema looks like what we need, we can use _id for character identification it makes more sense you are right. I am pretty new to all this. Would it make sense to include information about when the information was created updated?

for this schema we will need a write and read. If possible it would be nice to have a read function that has charID,startDate and endDate as parameters.

**Comment by kordianbruck** [comment]
No, dates don't really make sense in this case as the values get inserted once and thats it. Maybe a creation date, but you won't be updating the data will you?

So are you now directly accessing the DB or do you want an API? What did we agree on?

**Comment by kajo404** [comment]
I think we agreed on an API. The problem is that our twitter scrapers don't get 100% of the tweets as the twitter api limits the requests. Our data might actually change but we can just override the existing data then

**Comment by sacdallago** [comment]
On the last messages:
1. @kajo404 A can provide you with a POST request to update the sentiment of a character given a certain interval
2. @kordianbruck if they are gonna use the API to populate the sentiments you should go by name, if you are gonna use the same database and they will store a relation in their schema to the document of the character, you **can** use the _id, but being the name a unique key that's also not wrong, although speed is > on _id

**Comment by kajo404** [comment]
Just to update with the information we discussed offline on Friday after the presentation. We will use the character name. 

When do you think the API calls will be available for use? Again, we need to save data in the schema above and make calls preferably in a date range that gives us all the datasets for a character in the timeframe.

**Comment by kajo404** [comment]
Hi guys, really sorry to hit you with this so late but we need another API call. 

We need to be able to get all sentiments for all characters in a timeframe to calculate the top 3 in that timeframe. Otherwise we would need to make ~120 requests for this calculation :D

**Comment by kajo404** [comment]
## Nevermind I didn't read

 I see what you did there, as mentioned above we also want to save a date (we calculate the score for every day)

**Comment by togiberlin** [comment]
@kajo404 Hi Kajo, I think it's already in there (see line 6): https://github.com/Rostlab/JS16_ProjectA/blob/master/app/models/characterSentiment.js

**Comment by kajo404** [comment]
Yes I saw that, but we are supposed to access the schema with the rest api. What are the urls to do the requests?

**Comment by togiberlin** [comment]
@kajo yeah, I guess this needs to be done. Let me do that until March 23rd, evening

**Comment by jcebel** [comment]
Hey :) 
Does the urls already exist? I can't find them in the docs..

**Comment by togiberlin** [comment]
@jonny3576 Sorry for delay, please be patient. I'm also writing 2 retake exams soon. But I want to close this issue this evening!

**Comment by togiberlin** [comment]
@kajo404 @jonny3576 The basic stuff is there:
`var characterSentimentController = require(__appbase + 'controllers/characterSentiment');`
`router.post('/characters/sentiment/find', characterSentimentController.get);`
`router.get('/characters/sentiment', characterSentimentController.getAll);`
`router.get('/characters/sentiment/:date', characterSentimentController.getByDate);`
`router.get('/characters/sentiment/:id', characterSentimentController.getById);`
`router.get('/characters/sentiment/:description', characterSentimentController.getByDescription);`
If you need more, put a comment below and mention me with @togiberlin

**Comment by kajo404** [comment]
Hello @togiberlin ,
I have several questions about this.
1. How do we insert data into the database?
2. Is it possible to get make requests so they return everything in a timeframe?
3. Is any other group actually using this schema? (is description used at all?)
4. Can you provide an example of how to insert data?

**Comment by jcebel** [comment]
@togiberlin could you at least give us some status? This very important for the Project, especially as feature freeze is coming up tomorrow...

**Comment by ghost** [comment]
Status?

Milestone: feature freeze...

@togiberlin @kajo404 @jonny3576 @kordianbruck @sacdallago

**Comment by jcebel** [comment]
Yeah I know.. D5 was waiting for the promised functionality. But I think it is now resolved... Is it @kajo404 ?

**Comment by kajo404** [comment]
I just copied some links from your routes.js . I don't know if the functionality is fully implemented

**Comment by ghost** [comment]
@togiberlin

**Comment by togiberlin** [comment]
@adiolis @kajo404  All GET methods should work. However, I couldn't figure out this access token thing, which is needed for adding/editing/removing data via API. I don't know if it's a bug or something. I asked @kordianbruck over fb for assistance. I think he knows better than me, how this access token works as he implemented this security feature

**Comment by sacdallago** [comment]
Seriously :D Are we speaking about _that_ token? Guys?!?!?!?!?!

**Comment by togiberlin** [comment]
Actually what you do is append ?token=XYZ123 to the POST request.  But Postman is returning error message "accessToken not defined". I wasn't able to figure it out.

**Comment by sacdallago** [comment]
Are you sure it's urlencoded and not part of the head or body? @kordianbruck changed the token, as far as I remember from a conversation yesterday

**Comment by togiberlin** [comment]
@sacdallago I explained already to @kordianbruck via fb, he said he will look into it

**Comment by kajo404** [comment]
Can I get an update on the token situation? I also get the token is undefined error

**Comment by ghost** [comment]
@kordianbruck We need him to response :stuck_out_tongue_closed_eyes:

**Comment by ghost** [comment]
@togiberlin So, just testing is remaining?

**Comment by kordianbruck** [comment]
All updated and token working now.

**Comment by sacdallago** [comment]
What was the problem?
Can this issue be closed?

**Comment by kajo404** [comment]
I want to test my code before I close it ;) Give me a couple of hours as I have quite a few bugs to fix

**Comment by kajo404** [comment]
I have one more issue. When I try to save a character I need to use the character name as a identifying "object"(this was decided during presentation week) When I try to save an entry with `'character' : "Jon Snow"` which should be a viable character name accoring to your character list I get the following error:

`{"message":"Error.","error":{"message":"CharacterSentiment validation failed","name":"ValidationError","errors":{"character":{"message":"Cast to ObjectID failed for value \"Jon Snow\" at path \"character\"","name":"CastError","kind":"ObjectID","value":"Jon Snow","path":"character"}}}}
`

**Comment by ghost** [comment]
@kajo404 Fixed it. @kordianbruck Pull

**Comment by kajo404** [comment]
I am still getting the same error, do I need to wait for @kordianbruck to work some magic?

**Comment by sacdallago** [comment]
Yeah he needs to pull the last version to his server. He will write here once that is done, patience.
Otherwise you can pull the last develop of **A** and ask them for the mlab credentials ;)

**Comment by kordianbruck** [comment]
New code is live. Character collection is updating now.

**Comment by kordianbruck** [comment]
@sacdallago I guess this can be closed as its also live on the new server?

**Comment by sacdallago** [comment]
Yup.
